packages/modeling/src/operations/modifiers/retessellate.js

```diff
@@ -26,29 +18,77 @@ const retessellate = (geometry) => {
-    if (mapping) {
-      const polygons = mapping[1]
-      polygons.push(polygon)
+  const polygons = geom3.toPolygons(geometry).map((polygon, index) => ({vertices: polygon.vertices, plane: poly3.plane(polygon), index: index}))
+  const classified = classifyPolygons(polygons)
+
+  const destPolygons = []
+  classified.forEach((group) => {
+    if (Array.isArray(group)) {
+      const reTessellateCoplanarPolygons = reTesselateCoplanarPolygons(group)
+      destPolygons.push(...reTessellateCoplanarPolygons)
     } else {
-      polygonsPerPlane.push([poly3.plane(polygon), [polygon]])
+      destPolygons.push(group)
     }
   })
 
-  let destpolygons = []
-  polygonsPerPlane.forEach((mapping) => {
-    const sourcepolygons = mapping[1]
-    const retesselayedpolygons = reTesselateCoplanarPolygons(sourcepolygons)
-    destpolygons = destpolygons.concat(retesselayedpolygons)
-  })
-
-  const result = geom3.create(destpolygons)
+  const result = geom3.create(destPolygons)
   result.isRetesselated = true
 
   return result
 }
 
-module.exports = retessellate
+const classifyPolygons = (polygons) => {
+  let clusters = [polygons] // a cluster is an array of potentially coplanar polygons
+  const nonCoplanar = [] // polygons that are known to be non-coplanar
+  // go through each component of the plane starting with the last one (the distance from origin)
+  for (let component = 3; component >= 0; component--) {
+    const maybeCoplanar = []
+    clusters.forEach(cluster => {
+      // sort the cluster by the current component
+      cluster.sort(byPlaneComponent(component))
+      // iterate through the cluster and check if there are polygons which are not coplanar with the others 
+      // or if there are sub-clusters of coplanar polygons
+      let startIndex = 0
+      for (let i = 1; i < cluster.length; i++) {
+        // if there's a difference larger than the tolerance, split the cluster
+        if (cluster[i].plane[component] - cluster[startIndex].plane[component] > tolerance) {
+          // if there's a single polygon it's definitely not coplanar with any others
+          if (i - startIndex == 1) {
+            nonCoplanar.push(cluster[startIndex])
+          } else { // we have a new sub cluster of potentially coplanar polygons
+            maybeCoplanar.push(cluster.slice(startIndex, i))
+          }
+          startIndex = i
+        }
+      }
+      // handle the last elements of the cluster
+      if (cluster.length - startIndex == 1) {
+        nonCoplanar.push(cluster[startIndex])
+      } else {
+        maybeCoplanar.push(cluster.slice(startIndex))
+      }
+    })
+    // replace previous clusters with the new ones
+    clusters = maybeCoplanar
+  }
+  // restore the original order of the polygons
+  const result = []
+  // polygons inside the cluster should already be sorted by index
+  clusters.forEach(cluster => result[cluster[0]?.index] = cluster)
```

Comment:
This is too cryptic. Can you expand the expression?

I think that the full function signature should also be provided. Same for the next line.

Reply:
I know it's a bit confusing, I'll try to explain the best way I can.
Before calling the `classifyPolygons()` function, on line 19, all polygons are remapped to contain vertices, the plane object and the original index.
The original index is important because during classification the polygons get sorted based on the components of the plane object so they're out of order from the perspective of their index.
For geom3 objects this is not really important because the order of the polygons doesn't really matter.
geom2 is where things start to get wrong. Boolean operations for geom2 objects have a "special" behavior - they transform geom2 to fake geom3, do the boolean operation and then turn the result back to a geom2 object. In this scenario, if the polygons are out of order then the resulting points of the geom2 object will be out of order.
Now that I've explained why we need to sort the result based on index, let's take a look at those lines of code you highlighted.
After classification we have two resulting arrays (`nonCoplanar` and `clusters`) which need to be put back together in a single array (`result`), ordered by index. 
Let's start with `nonCoplanar` as it's the easiest to explain. This is an array of polygons which are not coplanar with any others.
Line 77 iterates through each polygon and places it in the `result` array at it's original index.
For `clusters` there's a similar process with some small differences. `clusters` is an array of arrays of polygons. We iterate though each cluster and place it in the `result` array at the index of it's first polygon.
To make it even easier to understand, let's take a look at a small example. Let's assume we have a geometry with 10 polygons, some of which are coplanar. These would be the indexes of the polygons before classification `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`. At the end of the classification process we end up with something like this: `nonCoplanar`: `[9, 4]` and `clusters`: `[ [2, 7], [0, 5], [1, 6], [3, 8] ]`.
When putting things back to the `result` array we would have: 
```
result[9] = {index: 9, vertices: [...], plane: [...]}
result[4] = {index: 4, vertices: [...], plane: [...]}
result[2] = [{index: 2, vertices: [...], plane: [...]}, {index: 7, vertices: [...], plane: [...]}]
result[0] = [{index: 0, vertices: [...], plane: [...]}, {index: 5, vertices: [...], plane: [...]}]
result[1] = [{index: 1, vertices: [...], plane: [...]}, {index: 6, vertices: [...], plane: [...]}]
result[3] = [{index: 3, vertices: [...], plane: [...]}, {index: 8, vertices: [...], plane: [...]}]
```
This is exactly the same behavior as the previous implementation.
Sorry for the wall of text, hope it helps.

In terms of providing full function signature - are you suggesting I should extract these 2 lines in separate methods?

Reply:
@achirita got it. I understand the sorting / combing into grouped.

it not about that... the planes are normals plus distance from origin. Grouping distances is not the same as grouping normals . The tolerances are different.

Reply:
I just updated that based on your comments